### PR TITLE
fix: correct IP selection logic

### PR DESCRIPTION
**Pull request by Kxuan**
`pickIP` reorder the sorted addresses using the user configured `prioritySubnets`.

However, if no address hits the `prioritySubnets`, the weights for all addresses is `0`. In this case, due to the uncertainty of map iteration order in Go, an address will be selected randomly.

https://github.com/Zebradil/cloudflare-dynamic-dns/blob/7a9965961c7211ce3b8fa31ffaf9dabf4f378b40/cmd/ip.go#L149-L156

For example, if there are two address in `ips`. The one is GUA address, and another one is ULA address. If the user does not have `prioritySubnets` set, all the two addresses will have weight `maxWeight`. The code above will select the first address picked by Go, not GUA.

This commit fix this issue by replace the whole address selection code with a score algorithm.

The new scoring algorithm uses the existing selection logic in the document. It gives a score according to the value of address. More important factors will affect the higher bits of the score.

So address sorting is no longer ignored.

### CODE REVIEW

File: cmd/ip.go [hunk 1]
```diff
@@ -73,89 +68,86 @@ func (s ipManager) getAllIPs() []net.IP {
...
+	for order, subnet := range s.cfg.prioritySubnets {
 		_, ipNet, err := net.ParseCIDR(subnet)
 		if err != nil {
 			log.WithError(err).WithField("subnet", subnet).Error("Couldn't parse subnet")
 			continue
 		}
-		netPrioritySubnets = append(netPrioritySubnets, *ipNet)
+		order++
+		if ipNet.Contains(ip) {
+			maxOrder = order
```
Comment (Zebradil) [inline cmd/ip.go:124]: We should stop checking priority subnets after the first match. A subnet is more preferable, higher it is in the list. If we continue and if the IP matches more than on subnet, it'll get higher weight.

```suggestion
			maxOrder = order
			break
```

File: cmd/ip.go [hunk 2]
```diff
@@ -73,89 +68,86 @@ func (s ipManager) getAllIPs() []net.IP {
...
+		if ipv4IsSAS(ip) {
+			score += 0x4
+		}
+		if ip.IsPrivate() {
+			score += 0x2
+		}
+	} else {
+		return 0
 	}
+	score += 0x1
```
Comment (Zebradil) [inline cmd/ip.go:112]: This line is a little confusing. Let's initialize `score` with `0x1` from the start.

**Review by Zebradil — CHANGES_REQUESTED** [review]
Hi @Kxuan,
Thanks a lot for your contribution and detailed explanation of the issue. I really like how elegant the scoring approach is.

I've left a few comments. Other than that the PR looks good 👍

**Comment by Kxuan** [comment]
Hi @Zebradil . Thanks for your review. I have applied your comment, and updated the branch. Please take a look.

In addition, I also moved the code about address types into `ipv4Stack` and `ipv6Stack`. The old code in my patch mistakenly ignores `cfg.stack`.

File: cmd/ip.go [hunk 3]
```diff
@@ -73,89 +69,89 @@ func (s ipManager) getAllIPs() []net.IP {
...
-	weightedAddresses := make(map[string]int)
-	for _, ip := range ips {
-		weightedAddresses[ip.String()] = maxWeight
-		for i, ipNet := range netPrioritySubnets {
-			if ipNet.Contains(ip) {
-				weightedAddresses[ip.String()] = i
-				break
-			}
+		if ipNet.Contains(ip) {
+			score += uint64((1 + len(s.cfg.prioritySubnets) - order)) << 16
```
Comment (Zebradil) [inline cmd/ip.go:130]: I think the `+1` is not needed here anymore.

```suggestion
			score += uint64(len(s.cfg.prioritySubnets) - order) << 16
```

File: cmd/ip.go [hunk 4]
```diff
@@ -73,89 +69,89 @@ func (s ipManager) getAllIPs() []net.IP {
...
-	ipv4s := []net.IP{}
-	for _, ip := range ips {
-		if ip.To4() != nil {
-			ipv4s = append(ipv4s, ip)
+	if ipv6IsGUA(ip) {
+		score += 0x8
+		if ipv6IsEUI64(ip) {
+			score += 0x4
 		}
 	}
```
Comment (Zebradil) [inline cmd/ip.go:87]: From [RFC-4193](https://datatracker.ietf.org/doc/html/rfc4193#section-3.2.2) and [RFC-4291](https://datatracker.ietf.org/doc/html/rfc4291#section-2.5.1), it seems like both Global Unicast addresses (GUA) and Unique Local addresses (ULA) can have the interface ID part in the modified EUI-64 format. So these score points do not depend on each other.

```suggestion
	if ipv6IsGUA(ip) {
		score += 0x8
	}

	if ipv6IsEUI64(ip) {
		score += 0x4
	}
```

**Review by Zebradil — CHANGES_REQUESTED** [review]

**Comment by Kxuan** [comment]
@Zebradil Sorry for the late update. I have modified the patch, please take a look again. Thanks.

**Comment by Zebradil** [comment]
Thanks! I think we're good to go. Did you test it in some way? Unfortunately, I haven't implemented any automated ways to test changes, so this has to be done manually. I'll do some test before merging, but this might take some time.

**Review by Zebradil — APPROVED** [review]
There was a bug in a situation when IPs from a wrong stack were getting higher scores than IPs from another stack. I added a step of filtering by stack before scoring.

Apart from that, I moved code around and renamed a few things.

**Pull request merged**